isobus/src/isobus_heartbeat.cpp

```diff
@@ -0,0 +1,287 @@
+
+		if (nullptr != pgnRequestProtocol)
+		{
+			pgnRequestProtocol->register_request_for_repetition_rate_callback(static_cast<std::uint32_t>(CANLibParameterGroupNumber::HeartbeatMessage), process_request_for_heartbeat, this);
+		}
+	}
+
+	EventDispatcher<HeartbeatInterface::HeartBeatError, std::shared_ptr<ControlFunction>> &HeartbeatInterface::GetHeartbeatErrorEventDispatcher()
+	{
+		return heartbeatErrorEventDispatcher;
+	}
+
+	EventDispatcher<std::shared_ptr<ControlFunction>> &HeartbeatInterface::GetNewTrackedHeartbeatEventDispatcher()
+	{
+		return newTrackedHeartbeatEventDispatcher;
+	}
+
+	void HeartbeatInterface::update()
+	{
+		if (enabled && initialized)
+		{
+			trackedHeartbeats.erase(std::remove_if(trackedHeartbeats.begin(), trackedHeartbeats.end(), [this](Heartbeat &heartbeat) {
+				                        bool retVal = false;
+
+				                        if (nullptr != heartbeat.controlFunction)
+				                        {
+					                        if (ControlFunction::Type::Internal == heartbeat.controlFunction->get_type())
+					                        {
+						                        if ((SystemTiming::time_expired_ms(heartbeat.timestamp_ms, heartbeat.repetitionRate_ms)) &&
+						                            heartbeat.send())
+						                        {
+							                        heartbeat.sequenceCounter++;
+
+							                        if (heartbeat.sequenceCounter > 250)
+							                        {
+								                        heartbeat.sequenceCounter = 0;
+							                        }
+						                        }
+					                        }
+					                        else if (SystemTiming::time_expired_ms(heartbeat.timestamp_ms, SEQUENCE_TIMEOUT_MS))
+					                        {
+						                        retVal = true; // External heartbeat is timed-out
+						                        LOG_ERROR("[HB]: Heartbeat from control function at address 0x%02X timed out.", heartbeat.controlFunction->get_address());
+						                        heartbeatErrorEventDispatcher.call(HeartBeatError::TimedOut, heartbeat.controlFunction);
+					                        }
+				                        }
+				                        else
+				                        {
+					                        retVal = true; // Invalid state
+				                        }
+				                        return retVal;
+			                        }),
+			                        trackedHeartbeats.end());
+		}
+	}
+
+	HeartbeatInterface::Heartbeat::Heartbeat(std::shared_ptr<ControlFunction> sendingControlFunction) :
+	  controlFunction(sendingControlFunction),
+	  timestamp_ms(SystemTiming::get_timestamp_ms())
+	{
+	}
+
+	bool HeartbeatInterface::Heartbeat::send()
+	{
+		bool retVal = false;
+		const std::array<std::uint8_t, 1> buffer = { sequenceCounter };
+
+		retVal = CANNetworkManager::CANNetwork.send_can_message(static_cast<std::uint32_t>(CANLibParameterGroupNumber::HeartbeatMessage),
```

Comment:
resolving circular-dependency: this is the interesting part, to make this interface completely free from the network manager then we probably need a callback for sending messages like in the recently refactored transport protocols? I still don't really like that approach but I haven't found anything nicer yet

Reply:
Yeah, I dunno if I super love how we're doing the callback thing with those, but I agree I don't have an alterative to propose, so for the time being we might as well keep it consistent and do something similar...

Reply:
Seems you forgot to implement `sendCANFrameCallback` here?

Reply:
Quite right, fixed.